src/jvm/main/org/jetbrains/kotlinx/lincheck/Planner.kt

```diff
@@ -0,0 +1,254 @@
+            return
+        }
+        // if we over-perform by and order of magnitude, increase the number of invocations per iteration
+        if (adjustInvocations && remainingTime / timeEstimate > INVOCATIONS_FACTOR) {
+            invocationsBound = min(INVOCATIONS_UPPER_BOUND, invocationsBound * INVOCATIONS_FACTOR)
+            return
+        }
+
+        // if iteration number adjustment is disabled there is nothing else we can do
+        if (!adjustIterations)
+            return
+
+        // if we under-perform, decrease the number of planned iterations
+        if (remainingTime < timeEstimate) {
+            // how bad is situation?
+            val delay = timeEstimate - remainingTime
+            val delayingIterations = floor(delay / (invocationsBound * averageInvocationTime)).toInt()
+            // remove the iterations we are unlikely to reach
+            iterationsUpperBound -= delayingIterations
+            return
+        }
+        val invocationsDelta = ITERATIONS_DELTA * invocationsBound
+        // if we over-perform, schedule more iterations
+        if (remainingTime - timeEstimate > invocationsDelta * averageInvocationTime) {
+            iterationsUpperBound += ITERATIONS_DELTA
+            return
+        }
+    }
+
+    companion object {
+        private fun LincheckOptions.iterationsLowerBound() =
+            if (adjustIterations)
+                // if automatic adjustment is enabled, we definitely should run custom scenarios,
+                // other scenarios only if there will be time left
+                customScenarios.size
+            else
+                // if automatic adjustment is disabled, we have should perform
+                // exactly the following number of iterations (as prescribed by the user)
+                customScenarios.size + iterations
+
+        private fun LincheckOptions.iterationsUpperBound() =
+            if (adjustIterations)
+                max(iterationsLowerBound(), iterations)
+            else
+                // if automatic adjustment is disabled, upper bound is equal to lower bound,
+                // as we should perform the exact number of iterations
+                iterationsLowerBound()
+
+        private fun LincheckOptions.iterationsStrictBound() =
+            if (!adjustIterations) iterationsLowerBound() else null
+
+        // number of iterations added/subtracted when we over- or under-perform the plan
+        private const val ITERATIONS_DELTA = 5
+
+        // factor of invocations multiplied/divided by when we over- or under-perform the plan
+        // by an order of magnitude
+        private const val INVOCATIONS_FACTOR = 5
+
+        // lower bound of invocations allocated by iteration
+        private const val INVOCATIONS_LOWER_BOUND = 1_000
+
+        // upper bound of invocations allocated by iteration
+        private const val INVOCATIONS_UPPER_BOUND = 100_000
```

Comment:
Does it work well with the current model checker? I guess the tests will fail with `OutOfMemoryError`.